README
NAME
    WWW::HtmlUnit - Inline::Java based wrapper of the HtmlUnit v2.9 library

SYNOPSIS
      use WWW::HtmlUnit;
      my $webClient = WWW::HtmlUnit->new;
      my $page = $webClient->getPage("http://google.com/");
      my $f = $page->getFormByName('f');
      my $submit = $f->getInputByName("btnG");
      my $query  = $f->getInputByName("q");
      $page = $query->type("HtmlUnit");
      $page = $query->type("\n");

      my $content = $page->asXml;
      print "Result:\n$content\n\n";

DESCRIPTION
    This is a wrapper around the HtmlUnit library (HtmlUnit version
    2.9-SNAPSHOT (2011.06.05) for this release). It includes the HtmlUnit
    jar itself and it's dependencies. All this library really does is find
    the jars and load them up using Inline::Java.

    The reason all this is interesting? HtmlUnit has very good javascript
    support, so you can automate, scrape, or test javascript-required
    websites.

    See especially the HtmlUnit documentation on their site for deeper API
    documentation, <http://htmlunit.sourceforge.net/apidocs/>.

INSTALLING
    There is one special thing that I've run into when installing
    Inline::Java, and thus WWW::HtmlUnit, which is telling the installer
    where to find your java home. It turns out this is really really easy,
    just define the JAVA_HOME environment variable before you start your
    CPAN shell / installer. From Debian/Ubuntu, I do:

      sudo apt-get install default-jdk
      sudo JAVA_HOME=/usr/lib/jvm/default-java cpanm WWW::HtmlUnit

    and everything works the way I want!

DOCUMENTATION
    You can get the bulk of the documentation directly from the HtmlUnit
    apidoc site <http://htmlunit.sourceforge.net/apidocs/>. Since
    WWW::HtmlUnit is mostly a wrapper around the real Java API, what you
    actually have to do is translate some of the java notation into perl
    notation. Mostly this is replacing '.' with '->'.

    Key classes that you might want to look at:

    WebClient
    <http://htmlunit.sourceforge.net/apidocs/com/gargoylesoftware/htmlunit/W
    ebClient.html>
        Represents a web browser. This is what "WWW::HtmlUnit->new" returns.

    HtmlPage
    <http://htmlunit.sourceforge.net/apidocs/com/gargoylesoftware/htmlunit/h
    tml/HtmlPage.html>
        A single HTML Page.

    HtmlElement
    <http://htmlunit.sourceforge.net/apidocs/com/gargoylesoftware/htmlunit/h
    tml/HtmlElement.html>
        An individual HTML element (node).

    Also see WWW::HtmlUnit::Sweet for a way to pretend that HtmlUnit works a
    little like WWW::Mechanize, but not really.

MODULE IMPORT PARAMETERS
    In general, any parameters you pass while importing ('use'-ing)
    WWW::HtmlUnit will be passed on to Inline::Java. A handy one is the
    'DIRECTORY' parameter, for example. A few parameters are handled
    specially, however.

    If you need to include extra .jar files, and/or if you want to study
    more java classes, you can do:

      use HtmlUnit
        jars => ['/path/to/blah.jar'],
        study => ['class.to.study'];

    and that will be added to the list of jars for Inline::Java to
    autostudy, and add to the list of classes for Inline::Java to
    immediately study. A class must be on the study list to be directly
    instantiated.

    Whether you ask for it or not, WebClient, BrowserVersion, and Cookie
    (each in the com.gargoylesoftware.htmlunit package) are studied. You can
    get to studied classes by adding WWW::HtmlUnit:: to their package name.
    So, you could make a cookie like this:

      my $cookie = WWW::HtmlUnit::com::gargoylesoftware::htmlunit::Cookie->new($name, $value);
      $webClient->getCookieManager->addCookie($cookie);

    Which is, incidentally, just the sort of thing that I should wrap in
    WWW::HtmlUnit::Sweet or elsewhere, 'cause that is UGLY!

METHODS
  $webClient = WWW::HtmlUnit->new($browser_name)
    This is just a shortcut for

      $webClient = WWW::HtmlUnit::com::gargoylesoftware::htmlunit::WebClient->new;

    The optional $browser_name allows you to specify which browser version
    to pass to the WebClient->new method. You could pass "FIREFOX_3" for
    example, to make the engine especially try to emulate Firefox 3 quirks,
    I imagine.

DEPENDENCIES
    When installed using the CPAN shell, all dependencies besides java
    itself will be installed. This includes the HtmlUnit jar files, and in
    fact those files make up the bulk of the distribution, byte-wise.

TIPS
  Working with java list/collections
    When you get a java list, it is actually an object-thingie. You gotta
    call "->toArray()" on it, and then you'll get a lovely perl arrayref,
    which is most likely what you wanted in the first place. I am open to
    suggestions for a mass work-around for this.

  HTTP Authentication
      my $credentialsProvider = $webclient->getCredentialsProvider;                           
      $credentialsProvider->addCredentials($username, $password);

  Disable SSL certificate checking
      $webclient->setUseInsecureSSL(1);

  Handling alerts and confirmations
    We (thanks lungching!) wrote a wee bit of java to make this easy. Though
    I admit that it could be a bit more... perlish. For a full example, see
    "03_clickhandler.t" in t.

      my $alert_handler = WWW::HtmlUnit::com::gargoylesoftware::htmlunit::CollectingAlertHandler->new();
      $webClient->setAlertHandler($alert_handler);
      # ...
      my $alert_arrayref = $alert_handler->getCollectedAlerts->toArray();

TODO
    *   Capture HtmlUnit output to a variable

    *   Use that to have a quiet-mode

    *   Document lungching's confirmation handler code, automate build

SEE ALSO
    WWW::HtmlUnit::Sweet, <http://htmlunit.sourceforge.net/>, Inline::Java

AUTHOR
      Brock Wilcox <awwaiid@thelackthereof.org> - http://thelackthereof.org/

COPYRIGHT
      Copyright (c) 2009-2011 Brock Wilcox <awwaiid@thelackthereof.org>. All rights
      reserved.  This program is free software; you can redistribute it and/or
      modify it under the same terms as Perl itself.

      HtmlUnit library includes the following copyright:

        /*
         * Copyright (c) 2002-2010 Gargoyle Software Inc.
         *
         * Licensed under the Apache License, Version 2.0 (the "License");
         * you may not use this file except in compliance with the License.
         * You may obtain a copy of the License at
         * http://www.apache.org/licenses/LICENSE-2.0
         *
         * Unless required by applicable law or agreed to in writing, software
         * distributed under the License is distributed on an "AS IS" BASIS,
         * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
         * See the License for the specific language governing permissions and
         * limitations under the License.
         */



Changes
Revision history for Perl extension WWW::HtmlUnit.

0.21 2013.07.25
  - Update to HtmlUnit 2.12

0.20 2012.12.18
  - Update to HtmlUnit 2.11

0.19 2012.12.18
  - Er... fix classpath better
  - Fix manifest

0.18 2012.11.14
  - Fix classpath separator

0.17 2012.10.23
  - Update to HtmlUnit 2.10

0.16 2011.06.06
  - Update to HtmlUnit 2.9 (TONS of bug fixes!)
  - BEWARE: In WWW::HtmlUnit::Sweet, auto-call toArray

0.15 2011.02.28
  - Pass import (use) parameters through to Inline::Java
  - Add WWW::HtmlUnit::Sweet parameters for stderr handling

0.14 2011.01.12
  - Documentation updates (minor)
  - Fix a test
  - Bumped dependency on Inline::Java to 0.53

0.13 2010.12.03
  - Add clickhandler (thanks lungching!)

0.12 2010.11.25
  - Fix 'study' usage
  - Minor documentation fixes

0.11 2010.11.22
  - Oops, forgot to update MANIFEST

0.10 2010.11.22
  - Add WWW::HtmlUnit::Sweet
  - Add 'study' like 'jars' import param (Fred Castellano)
  - Add some more documentation
  - Note that HtmlUnit is now up to 2.8

0.07 2009.09.03.01.36
  - Update to HtmlUnit 2.6

0.06 2009.09.03.01.26
  - Add jars parameter (thanks Martin Drashkov!)
  - Update documentation

0.05 2009.05.16.19.40
  - Fix jar search path a bit

0.04 2009.04.24.23.44
  - Update to HtmlUnit 2.5
  - Move into the WWW namespace

0.03 2008.12.31.04.40
  - Update to HtmlUnit 2.4
  - Include google calendar .ics file upload example

0.02 2008.12.30.14.46
  - Fixed module description
  - Fixed the MANIFEST to actually include the jar files

0.01 2008.12.30.01.45
	- Initial version



lib/WWW/HtmlUnit/Sweet.pm
package WWW::HtmlUnit::Sweet;

=head1 NAME

WWW::HtmlUnit::Sweet - Wrapper around WWW::HtmlUnit to add some sweetness

=head1 SYNOPSIS

  use WWW::HtmlUnit::Sweet;
  my $agent = WWW::HtmlUnit::Sweet->new;

  $agent->getPage('http://google.com/');

  # Type into the currently focused element
  $agent->type("Hello\n");

  # Print out the XML of the page
  print $agent->asXml;

=head1 DESCRIPTION

Using L<WWW::HtmlUnit> as a foundation, this adds some convenience things. The main addition is that the $agent you get from ->new does some AUTOLOAD things to allow you to treat the $agent as either a browser, a window, or a page. That way you can treat it a lot more like a L<WWW::Mechanize> object.

This module might change drastically, buyer beware!

=head1 IMPORT PARAMETERS

When you 'use' this module, you can pass some parameters. Any parameter that L<WWW::HtmlUnit::Sweet> doesn't use will be passed on to L<WWW::HtmlUnit>, or ultimately L<Inline::Java>.

=over 4

=item * show_errors - Flag to stop the supression of stderr

=item * error_filename - Filename to append stderr to

=item * error_fh - Filehandle to append stderr to

=item * errors_to_tmpfile - Send stderr to a temporary file (L<IO::File>)

=back 4

Useful examples:

  # Show errors on STDERR
  use WWW::HtmlUnit::Sweet show_errors => 1;

  # Append errors to /tmp/errors.txt
  use WWW::HtmlUnit::Sweet error_filename => '/tmp/errors.txt';

Note that if you don't pass anything, errors will be sent to /dev/null (or a temporary file if you don't have /dev/null).

=cut

use strict;
use warnings;

# Hold our error filehandle
our $error_fh;

sub import {
  my $class = shift;
  my %parameters = @_;

  if($parameters{show_errors}) {
    delete $parameters{show_errors};
    require WWW::HtmlUnit;
    WWW::HtmlUnit->import( %parameters );
  } else {
    if($parameters{error_filename}) {
      open $error_fh, '>>', $parameters{error_filename}
        or die "Error opening $parameters{error_filename}, $!\n";
      delete $parameters{error_filename};
    } elsif($parameters{error_fh}) {
      $error_fh = $parameters{error_fh};
      delete $parameters{error_fh};
    } elsif($parameters{errors_to_tmpfile} || ! -c '/dev/null') {
      require IO::File;
      $error_fh = IO::File->new_tmpfile;
      delete $parameters{errors_to_tmpfile};
    } else {
      open $error_fh, '>', '/dev/null'
        or die "Error opening $parameters{error_filename}, $!\n";
    }

    # So we save STDERR, then redirect it

    no warnings; # stop complaint about SAVEERR never being used again
    open SAVEERR, '>&', STDERR;
    use warnings;
    close STDERR;
    open STDERR, '>&', $error_fh;
 
    # Now Inline::Java will use our special filehandle instead of STDERR
    require WWW::HtmlUnit;
    WWW::HtmlUnit->import( %parameters );

    # Now put STDERR back!
    close STDERR;
    open STDERR, '>&', SAVEERR;
	}

}

=head1 METHODS

=head2 $agent = WWW::HtmlUnit::Sweet->new

Create a new sweet agent. Use this kinda like looking at a browser on the screen. The methods you call will be invoked (if possible) on the current browser, window, page, or focused element.

The 'new' method can also take a browser version and a starting url, like this:

  my $agent = WWW::HtmlUnit::Sweet->new(
    version => 'FIREFOX_3',
    url => 'http://google.com/'
  );

=cut

sub new {
	my $class = shift;
	my $self = { @_ };
	bless $self, $class;
	$self->{browser} = WWW::HtmlUnit->new( $self->{version} );
  $self->getPage( $self->{url} ) if $self->{url};
	return $self;
}

=head2 $agent->wait_for(sub { ... }, $timeout)

Execute the provided sub once a second until it returns true, or until the the timeout has been reached. If a timeout isn't passed, it will default to 10 seconds (which you can change by setting C<< $WWW::HtmlUnit::Sweet::default_timeout >>). This is handy for waiting for the page to finish executing some javascript, or loading.

Example:

  # Wait for an element with id 'foo' to exist
  $agent->wait_for(sub {
    $agent->getElementById('foo')
  });

=cut

our $default_timeout = 10;

sub wait_for {
  my ($agent, $subref, $timeout) = @_;
  $timeout ||= $default_timeout;
  while($timeout) {
    return if eval { $subref->() } && ! $@;
    sleep 1;
    $timeout--;
  }
  die "Timeout!\n";
}

=head2 AUTOLOAD, aka $agent->whatever(..)

This is where the sweetness starts kicking in. First it will try to call ->whatever on the browser, and if there is no method named 'whatever' there it will be called on the current window, and if there is no method named 'whatever' there it will be called on the current page in that window, and if there is no method 'whatever' there it will be called on the currently focused element.

Examples:

  # This works at the browser level
  $agent->getPage('http://google.com/');

  # Get the 'name' for the current window
  my $window_name = $agent->getName;

  # Working from the current page, get an element by ID
  my $sidebar_element = $agent->getElementById('sidebar');

  # Click on the currently focused element
  $agent->click;

This scheme works quite well because HtmlUnit itself just so happens to not overlap their method names between different classes. Lucky us!

Note: We also call ->toArray on results if needed. Probably at some point we'll get ALL array-like results from HtmlUnit to auto-execute ->toArray.

=cut

# This will make us act a bit more like Mechanize
sub AUTOLOAD {
	my $self = shift;
	our $AUTOLOAD;
	my $method = $AUTOLOAD; $method =~ s/.*:://;
	return if $method eq 'DESTROY';
	my $retval = eval {
    
    my $browser = $self->{browser};
    my $window = $browser && $browser->getCurrentWindow;
    my $page = $window && $window->getEnclosedPage;
    my $element = $page && $page->getFocusedElement;

    my $result;
		if($browser && $browser->can($method)) {
			$result = $browser->$method(@_);
		} elsif($window && $window->can($method)) {
			$result = $window->$method(@_);
		} elsif($page && $page->can($method)) {
			$result = $page->$method(@_);
		} elsif($element && $element->can($method)) {
			$result = $element->$method(@_);
		} else {
			die "Method $method not found!";
		}
    if(ref $result && $result->can('toArray')) {
      return $result->toArray;
    } else {
      return $result;
    }
	};
	if($@ && ref($@) =~ /Exception/) {
		print STDERR "HtmlUnit ERROR: " . $@->getMessage . "\n";
		die $@; # Pass it up the chain
	} elsif($@) {
		warn $@;
	}
	return $retval;
}


package WWW::HtmlUnit::com::gargoylesoftware::htmlunit::html::HtmlSelect;

# Fix the get_option to take nicer params
# TODO: document this!

sub get_option {
	my ($self, %params) = @_;
	if($params{text}) {
		return eval {$self->getOptionByText($params{text})};
	} elsif($params{value}) {
		return eval {$self->getOptionByValue($params{value})};
	}
	die "Must pass either text or value";
}

=head1 TODO

Add more documentation and examples and sweetness :)

=head1 SEE ALSO

L<WWW::HtmlUnit>

=head1 AUTHOR

  Brock Wilcox <awwaiid@thelackthereof.org> - http://thelackthereof.org/

=head1 COPYRIGHT

  Copyright (c) 2009-2011 Brock Wilcox <awwaiid@thelackthereof.org>. All rights
  reserved.  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut

1;



eg/google_calendar_upload.pl
#!/usr/local/bin/perl

use strict;
use WWW::HtmlUnit;

# Put in your username and password, and then pass this script an .ics file.
# It'll upload it into your default google calendar

my $google_username = 'XXXXXXX';
my $google_password = 'XXXXXXX';

sub wait_for(&@) {
  my ($subref, $timeout) = @_;
  $timeout ||= 30;
  while($timeout) {
    return if eval { $subref->() };
    sleep 1;
    $timeout--;
  }
  die "Timeout!\n";
}

my $ics_filename = shift @ARGV;
if($ics_filename !~ /\.ics$/i) {
  print "Expected: google_calendar_upload.pl <filename.ics>\n";
  exit;
}

eval {
my $webClient = WWW::HtmlUnit->new('FIREFOX_3');
my $page = $webClient->getPage("http://google.com/calendar/");

my $login_form = $page->getElementById('gaia_loginform');
my $email = $login_form->getInputByName('Email');
my $passwd = $login_form->getInputByName('Passwd');
my $sign_in_button = $login_form->getInputByName('signIn');

$email->type($google_username);
$passwd->type($google_password);
$page = $sign_in_button->click;

wait_for { $page->getElementById('add_cals_link') };

$page->getElementById('add_cals_link')->click;

wait_for { $page->getElementById('addP') };

$page->getElementById('addP')->getLastChild->click;

wait_for { defined $page->getFrames->toArray->[1] };

my $p3 = $page->getFrames->toArray->[1]->getEnclosedPage;
my $f = $p3->getForms->toArray->[0];
# Had issues with the path of the form action
$f->setActionAttribute("http://www.google.com/calendar/" . $f->getActionAttribute);
my $filename = $f->getInputByName('filename');
$filename->setValueAttribute($ics_filename);
my $import_button = $p3->getByXPath('//*[@value="Import"]')->toArray->[0];
my $p4 = $import_button->click;
print $p4->asText;
print "\n";

};

if($@ && ref($@) =~ /Exception/) {
  print "Exception: " . $@->getMessage . "\n";
} elsif($@) {
  print "Err... $@\n";
}





lib/WWW/HtmlUnit.pm
package WWW::HtmlUnit;

=head1 NAME

WWW::HtmlUnit - Inline::Java based wrapper of the HtmlUnit v2.12 library

=head1 SYNOPSIS

  use WWW::HtmlUnit;
  my $webClient = WWW::HtmlUnit->new;
  my $page = $webClient->getPage("http://google.com/");
  my $f = $page->getFormByName('f');
  my $submit = $f->getInputByName("btnG");
  my $query  = $f->getInputByName("q");
  $page = $query->type("HtmlUnit");
  $page = $query->type("\n");

  my $content = $page->asXml;
  print "Result:\n$content\n\n";

=head1 DESCRIPTION

This is a wrapper around the HtmlUnit library. It includes the HtmlUnit jar itself and it's dependencies. All this library really does is find the jars and load them up using L<Inline::Java>.

The reason all this is interesting? HtmlUnit has very good javascript support, so you can automate, scrape, or test javascript-required websites.

See especially the HtmlUnit documentation on their site for deeper API documentation, L<http://htmlunit.sourceforge.net/apidocs/>.

=head1 INSTALLING

There is one special thing that I've run into when installing L<Inline::Java>, and thus L<WWW::HtmlUnit>, which is telling the installer where to find your java home. It turns out this is really really easy, just define the JAVA_HOME environment variable before you start your CPAN shell / installer. From Debian/Ubuntu, I do:

  sudo apt-get install default-jdk
  sudo JAVA_HOME=/usr/lib/jvm/default-java cpanm WWW::HtmlUnit

and everything works the way I want!

=head1 DOCUMENTATION

You can get the bulk of the documentation directly from the L<HtmlUnit apidoc site|http://htmlunit.sourceforge.net/apidocs/>. Since WWW::HtmlUnit is mostly a wrapper around the real Java API, what you actually have to do is translate some of the java notation into perl notation. Mostly this is replacing '.' with '->'.

Key classes that you might want to look at:

=over 4

=item L<WebClient|http://htmlunit.sourceforge.net/apidocs/com/gargoylesoftware/htmlunit/WebClient.html>

Represents a web browser. This is what C<< WWW::HtmlUnit->new >> returns.

=item L<HtmlPage|http://htmlunit.sourceforge.net/apidocs/com/gargoylesoftware/htmlunit/html/HtmlPage.html>

A single HTML Page.

=item L<HtmlElement|http://htmlunit.sourceforge.net/apidocs/com/gargoylesoftware/htmlunit/html/HtmlElement.html>

An individual HTML element (node).

=back

Also see L<WWW::HtmlUnit::Sweet> for a way to pretend that HtmlUnit works a little like L<WWW::Mechanize>, but not really.

=cut

use strict;
use warnings;

our $VERSION = '0.21';

sub find_jar_path {
  my $self = shift;
  my $path = $INC{'WWW/HtmlUnit.pm'};
  $path =~ s/\.pm$/\/jar/;
  return $path;
}

our $classpath_separator = $^O =~ /win/i ? ";" : ":";
sub collect_default_jars {
  my $jar_path = find_jar_path();
  return join $classpath_separator, map { "$jar_path/$_" } qw(
    commons-codec-1.7.jar
    commons-collections-3.2.1.jar
    commons-io-2.4.jar
    commons-lang3-3.1.jar
    commons-logging-1.1.1.jar
    cssparser-0.9.9.jar
    htmlunit-2.12.jar
    htmlunit-confirmhandler-2.8.jar
    htmlunit-core-js-2.12.jar
    httpclient-4.2.3.jar
    httpcore-4.2.2.jar
    httpmime-4.2.3.jar
    jetty-http-8.1.9.v20130131.jar
    jetty-io-8.1.9.v20130131.jar
    jetty-util-8.1.9.v20130131.jar
    jetty-websocket-8.1.9.v20130131.jar
    nekohtml-1.9.18.jar
    sac-1.3.jar
    serializer-2.7.1.jar
    xalan-2.7.1.jar
    xercesImpl-2.10.0.jar
    xml-apis-1.4.01.jar
  );
}

=head1 MODULE IMPORT PARAMETERS

In general, any parameters you pass while importing ('use'-ing) L<WWW::HtmlUnit> will be passed on to L<Inline::Java>. A handy one is the 'DIRECTORY' parameter, for example. A few parameters are handled specially, however.

If you need to include extra .jar files, and/or if you want to study more java classes, you can do:

  use HtmlUnit
    jars => ['/path/to/blah.jar'],
    study => ['class.to.study'];

and that will be added to the list of jars for L<Inline::Java> to autostudy, and add to the list of classes for L<Inline::Java> to immediately study. A class must be on the study list to be directly instantiated.

Whether you ask for it or not, WebClient, BrowserVersion, and Cookie (each in the com.gargoylesoftware.htmlunit package) are studied. You can get to studied classes by adding WWW::HtmlUnit:: to their package name. So, you could make a cookie like this:

  my $cookie = WWW::HtmlUnit::com::gargoylesoftware::htmlunit::Cookie->new($name, $value);
  $webClient->getCookieManager->addCookie($cookie);

Which is, incidentally, just the sort of thing that I should wrap in WWW::HtmlUnit::Sweet or elsewhere, 'cause that is UGLY!

=cut

sub import {
  my $class = shift;
  my %parameters = @_;
  my $custom_jars = "";
  if ($parameters{'jars'}) {
      $custom_jars = join($classpath_separator, @{$parameters{'jars'}});
      delete $parameters{'jars'};
  }

  my @STUDY = (
      'com.gargoylesoftware.htmlunit.WebClient',
      'com.gargoylesoftware.htmlunit.BrowserVersion',
      'com.gargoylesoftware.htmlunit.util.Cookie',
      'com.gargoylesoftware.htmlunit.CollectingAlertHandler',
      'com.gargoylesoftware.htmlunit.ClickConfirmHandler',
  );    
  if ($parameters{'study'}) {
      push(@STUDY, @{$parameters{'study'}});
      delete $parameters{'study'};
  }

  require Inline;
  Inline->import(
    Java => 'STUDY',
    STUDY => \@STUDY,
    AUTOSTUDY => 1,
    CLASSPATH => collect_default_jars() . $classpath_separator . $custom_jars,
    %parameters
  );
}

=head1 METHODS

=head2 $webClient = WWW::HtmlUnit->new($browser_name)

This is just a shortcut for 

  $webClient = WWW::HtmlUnit::com::gargoylesoftware::htmlunit::WebClient->new;

The optional $browser_name allows you to specify which browser version to pass to the WebClient->new method. You could pass "FIREFOX_3" for example, to make the engine especially try to emulate Firefox 3 quirks, I imagine.

=cut

sub new {
  my ($class, $version) = @_;
  if($version) {
    my $browser_version = eval "\$WWW::HtmlUnit::com::gargoylesoftware::htmlunit::BrowserVersion::$version";
    return WWW::HtmlUnit::com::gargoylesoftware::htmlunit::WebClient->new($browser_version);
  } else {
    return WWW::HtmlUnit::com::gargoylesoftware::htmlunit::WebClient->new;
  }
}

=head1 DEPENDENCIES

When installed using the CPAN shell, all dependencies besides java itself will be installed. This includes the HtmlUnit jar files, and in fact those files make up the bulk of the distribution, byte-wise.

=head1 TIPS

=head2 Working with java list/collections

When you get a java list, it is actually an object-thingie. You gotta call C<< ->toArray() >> on it, and then you'll get a lovely perl arrayref, which is most likely what you wanted in the first place. I am open to suggestions for a mass work-around for this.


=head2 HTTP Authentication

  my $credentialsProvider = $webclient->getCredentialsProvider;                           
  $credentialsProvider->addCredentials($username, $password);                

=head2 Disable SSL certificate checking

  $webclient->setUseInsecureSSL(1);

=head2 Handling alerts and confirmations

We (thanks lungching!) wrote a wee bit of java to make this easy. Though I admit that it could be a bit more... perlish. For a full example, see L<t/03_clickhandler.t>.

  my $alert_handler = WWW::HtmlUnit::com::gargoylesoftware::htmlunit::CollectingAlertHandler->new();
  $webClient->setAlertHandler($alert_handler);
  # ...
  my $alert_arrayref = $alert_handler->getCollectedAlerts->toArray();

=head1 TODO

=over 4

=item * Capture HtmlUnit output to a variable

=item * Use that to have a quiet-mode

=item * Document lungching's confirmation handler code, automate build

=back

=head1 SEE ALSO

L<WWW::HtmlUnit::Sweet>, L<http://htmlunit.sourceforge.net/>, L<Inline::Java>

=head1 AUTHOR

  Brock Wilcox <awwaiid@thelackthereof.org> - http://thelackthereof.org/

=head1 COPYRIGHT

  Copyright (c) 2009-2012 Brock Wilcox <awwaiid@thelackthereof.org>. All rights
  reserved.  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

  HtmlUnit library includes the following copyright:

    /*
     * Copyright (c) 2002-2012 Gargoyle Software Inc.
     *
     * Licensed under the Apache License, Version 2.0 (the "License");
     * you may not use this file except in compliance with the License.
     * You may obtain a copy of the License at
     * http://www.apache.org/licenses/LICENSE-2.0
     *
     * Unless required by applicable law or agreed to in writing, software
     * distributed under the License is distributed on an "AS IS" BASIS,
     * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
     * See the License for the specific language governing permissions and
     * limitations under the License.
     */

=cut

1;

